Objects/frameobject.c

```diff
@@ -13,9 +13,348 @@
+        if (PyTuple_GET_ITEM(co->co_localsplusnames, i) == key ||
+                _PyUnicode_EQ(PyTuple_GET_ITEM(co->co_localsplusnames, i), key)) {
+            PyObject *value = framelocalproxy_getval(frame, co, i);
+            if (value != NULL) {
+                return Py_NewRef(value);
+            }
+        }
+    }
+
+    // Okay not in the fast locals, try extra locals
+
+    PyObject *extra = frame->f_extra_locals;
+    if (extra != NULL) {
+        PyObject *value = PyDict_GetItem(extra, key);
+        if (value != NULL) {
+            return Py_NewRef(value);
+        }
+    }
+
+    PyErr_Format(PyExc_KeyError, "local variable '%R' is not defined", key);
+    return NULL;
+}
+
+static int
+framelocalsproxy_setitem(PyObject *self, PyObject *key, PyObject *value)
+{
+    /* Merge locals into fast locals */
+    PyFrameObject* frame = ((PyFrameLocalsProxyObject*)self)->frame;
+    PyObject** fast = _PyFrame_GetLocalsArray(frame->f_frame);
+    PyCodeObject* co = PyFrame_GetCode(frame);
+
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
+        if (PyTuple_GET_ITEM(co->co_localsplusnames, i) == key ||
+                _PyUnicode_EQ(PyTuple_GET_ITEM(co->co_localsplusnames, i), key)) {
+            _PyLocals_Kind kind = _PyLocals_GetKind(co->co_localspluskinds, i);
+
+            PyObject *oldvalue = fast[i];
+            PyObject *cell = NULL;
+            if (kind == CO_FAST_FREE) {
+                // The cell was set when the frame was created from
+                // the function's closure.
+                assert(oldvalue != NULL && PyCell_Check(oldvalue));
+                cell = oldvalue;
+            } else if (kind & CO_FAST_CELL && oldvalue != NULL) {
+                if (PyCell_Check(oldvalue)) {
+                    cell = oldvalue;
+                }
+            }
+            if (cell != NULL) {
+                oldvalue = PyCell_GET(cell);
+                if (value != oldvalue) {
+                    PyCell_SET(cell, Py_XNewRef(value));
+                    Py_XDECREF(oldvalue);
+                }
+            } else if (value != oldvalue) {
+                if (value == NULL) {
+                    // Probably can't delete this, since the compiler's flow
+                    // analysis may have already "proven" that it exists here:
+                    const char *e = "assigning None to unbound local %R";
+                    if (PyErr_WarnFormat(PyExc_RuntimeWarning, 0, e, key)) {
+                        // It's okay if frame_obj is NULL, just try anyways:
+                        PyErr_WriteUnraisable((PyObject *)frame);
+                    }
+                    value = Py_NewRef(Py_None);
+                }
+                Py_XSETREF(fast[i], Py_NewRef(value));
+            }
+            Py_XDECREF(value);
+            return 0;
+        }
+    }
+
+    // Okay not in the fast locals, try extra locals
+
+    PyObject *extra = frame->f_extra_locals;
+
```

Comment:
This is the point the `f_extra_locals` dict could be lazily created in order to avoid creating it in as many situations as posssible.